procrastinate/worker.py

```diff
@@ -260,17 +259,23 @@ async def _process_job(self, context: job_context.JobContext):
             )
 
     async def _fetch_and_process_jobs(self):
-        """Keeps in fetching and processing jobs until there are no job left to process"""
-        while True:
-            await self.job_semaphore.acquire()
+        """Fetch and process jobs until there are no job left ready to be processed"""
+        while not self._stop_event.is_set():
+            acquire_sem_task = asyncio.create_task(self._job_semaphore.acquire())
+            await utils.wait_any(acquire_sem_task, self._stop_event.wait())
+            if self._stop_event.is_set():
+                if acquire_sem_task.done():
+                    self._job_semaphore.release()
```

Comment:
Yes, you are right, we could, the way it is implemented, not release the semaphore on stop and it will be fine. 
I thought about it at the time too but decided it err on the side of caution (part of the code could change in the future and the absence of correct release introduce a subtle bug).

> That said, I wonder if it would make to do a single try/finally with a single self._job_semaphore.release() and release in all the case where the job isn't running.

Good idea